### PR TITLE
reduce yawollet requests to api server

### PR DESCRIPTION
**Pull request by breuerfelix**

### CODE REVIEW

File: controllers/yawol-cloud-controller/targetcontroller/node_controller_test.go [hunk 1]
```diff
@@ -17,6 +18,105 @@ import (
 	. "github.com/onsi/gomega"
 )
 
+var _ = Describe("check controller-runtime predicate", func() {
+	nodeName := "node1"
+	conditionReady := v1.NodeCondition{
+
+		Type:               v1.NodeReady,
```
Comment (breuerfelix) [inline controllers/yawol-cloud-controller/targetcontroller/node_controller_test.go:25]:
```suggestion
	conditionReady := v1.NodeCondition{
		Type:               v1.NodeReady,
```

File: controllers/yawol-cloud-controller/targetcontroller/node_controller_test.go [hunk 2]
```diff
@@ -17,6 +18,105 @@ import (
...
+		Status:             v1.ConditionTrue,
+		LastHeartbeatTime:  metav1.Time{},
+		LastTransitionTime: metav1.Time{},
+		Reason:             "Ready",
+		Message:            "Ready",
+	}
+	conditionNotReady := v1.NodeCondition{
+
+		Type:               v1.NodeReady,
```
Comment (breuerfelix) [inline controllers/yawol-cloud-controller/targetcontroller/node_controller_test.go:34]:
```suggestion
	conditionNotReady := v1.NodeCondition{
		Type:               v1.NodeReady,
```

File: controllers/yawol-controller/loadbalancerset/loadbalancerset_controller.go [hunk 3]
```diff
@@ -175,39 +175,49 @@ func (r *LoadBalancerSetReconciler) reconcileStatus(
...
+			if string(c.Type) == string(helper.KeepalivedMaster) &&
+				string(c.Status) != string(helper.ConditionFalse) {
+				return machines[i]
 			}
-			return ctrl.Result{RequeueAfter: time.Second * 2}, nil
 		}
 	}
 
-	return ctrl.Result{}, nil
+	return machines[0]
```
Comment (breuerfelix) [inline controllers/yawol-controller/loadbalancerset/loadbalancerset_controller.go:220]: do we need to check length? could this function be called on an empty machine list?

File: controllers/yawol-controller/loadbalancerset/loadbalancerset_controller.go [hunk 4]
```diff
@@ -175,39 +175,49 @@ func (r *LoadBalancerSetReconciler) reconcileStatus(
...
 	}
 
-	if desiredReplicas < currentReplicas-deletedReplicas {
-		// Delete not ready machines first
-		if len(notReadyMachines) > 0 {
-			if err := r.deleteMachine(ctx, &notReadyMachines[0]); err != nil {
-				return ctrl.Result{}, err
-			}
-			return ctrl.Result{RequeueAfter: time.Second * 2}, nil
+	if len(machines) < len(machines)-deletedMachineCount {
```
Comment (breuerfelix) [inline controllers/yawol-controller/loadbalancerset/loadbalancerset_controller.go:188]: shouldnt this be something like: len(machines) > desiredreplicas ? i don't quite understand why the deletedmachinecount is important for this calculation

File: docs/components.md [hunk 5]
```diff
@@ -77,7 +77,18 @@ OpenStack instance where the yawollet and Envoy do the actual Load Balancing.
 #### **loadbalancerset-controller**
 
 * Creates/Deletes `LoadBalancerMachines` from `LoadbalancerSet`
-* Monitor `LoadBalancerMachine` status and recreates `LoadBalancerMachine` if node is unhealthy
+* Monitor `LoadBalancerMachine` status and recreates `LoadBalancerMachine` if LoadBalancer is unhealthy.
+
+> A `LoadBalancerMachine` will me marked as unready if any of the following criteria matches:
+> - no condition are set
+> - any LastHeartbeatTime of condition is older than 3 min
```
Comment (breuerfelix) [inline docs/components.md:84]:
```suggestion
> - at least one LastHeartbeatTime of any condition is older than 3 min
```

File: controllers/yawollet/loadbalancer_controller.go [hunk 6]
```diff
@@ -44,105 +49,154 @@ func (r *LoadBalancerReconciler) Reconcile(ctx context.Context, req ctrl.Request
...
 
 // SetupWithManager is used by kubebuilder to init the controller loop
 func (r *LoadBalancerReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&yawolv1beta1.LoadBalancer{}).
+		WithEventFilter(yawolletPredicate(r.LoadbalancerName)).
 		Complete(r)
 }
+
+func yawolletPredicate(loadbalancerName string) predicate.Predicate {
```
Comment (timebertt) [inline controllers/yawollet/loadbalancer_controller.go:177]: Optionally, we could replace this predicate with the following:
```go
predicate.And(
	predicate.NewPredicateFuncs(func(obj client.Object) bool { return obj.GetName() == r.LoadbalancerName }),
	predicate.Or(
		predicate.AnnotationChangedPredicate{},
		predicate.GenerationChangedPredicate{},
	),
)
```

**Review by timebertt — COMMENTED** [review]
A few suggestions left, otherwise it looks good :)

**Review by timebertt — COMMENTED** [review]
A few suggestions left, otherwise it looks good :)

**Review by einfachnuralex — APPROVED** [review]

**Pull request merged**